### PR TITLE
Group OTLP data points under same metric

### PR DESCRIPTION
**Pull request by lenin-jaganathan**
This PR improves the OTLP export by effectively leveraging OpenTelemetry Line Protocol to encode metrics. This will both reduce the network bytes as well as the memory allocation we do in the JVM. I have attached a sample of what is changing. This will provide huge gains since in a real-world environment where there will only be a limited set of metrics that differ only in the tags/attributes (usually 100's - 1000s of time series for the same metric name and we will avoid duplicate publishing of the metadata). This will partially resolve https://github.com/micrometer-metrics/micrometer/issues/4053 by optimizing the publishing and eliminating most of the semantic errors.

### What is not addressed?
1. When a metric name has multiple descriptions (this shouldn't be an ideal usage but the micrometer allows this), we consider this a separate metric and add it to a new metric. (OTLP suggestion is to prefer the longest string but with micrometer design, this will need a lot of changes at the core and how a metric is treated in micrometer).
2. When a metric name has multiple base units, we will consider that a separate metric. (OTLP suggestion is to convert the data if possible.)

However, the above 2 will be extreme cases where the same metric name will have these properties with differing descriptions or baseUnit. Addressing this will need to have a different approach to how we identify metrics in micrometers and this corner case can be continued to be resolved as part of the mentioned issue.

### Comparison before and after:
Let's assume 2 counters - counter{tag=key1}, counter{tag=key2}. The description of the counters is "Test counter" and the base unit is "byte".

<details>
<summary>DataExported before this change </summary>

```
Metric #0
Descriptor:
     -> Name: counter
     -> Description: Test counter
     -> Unit: bytes
     -> DataType: Sum
     -> IsMonotonic: true
     -> AggregationTemporality: Delta
NumberDataPoints #0
Data point attributes:
     -> key: Str(value2)
StartTimestamp: 2024-01-12 08:50:25 +0000 UTC
Timestamp: 2024-01-12 08:50:30 +0000 UTC
Value: 1.000000
Metric #1
Descriptor:
     -> Name: counter
     -> Description: Test counter
     -> Unit: bytes
     -> DataType: Sum
     -> IsMonotonic: true
     -> AggregationTemporality: Delta
NumberDataPoints #0
Data point attributes:
     -> key: Str(value1)
StartTimestamp: 2024-01-12 08:50:25 +0000 UTC
Timestamp: 2024-01-12 08:50:30 +0000 UTC
Value: 1.000000
```
</details>

<details>
<summary>DataExported after this change </summary>

```
Metric #0
Descriptor:
     -> Name: counter
     -> Description: Test counter
     -> Unit: bytes
     -> DataType: Sum
     -> IsMonotonic: true
     -> AggregationTemporality: Delta
NumberDataPoints #0
Data point attributes:
     -> key: Str(value2)
StartTimestamp: 2024-01-12 08:50:25 +0000 UTC
Timestamp: 2024-01-12 08:50:30 +0000 UTC
Value: 1.000000
NumberDataPoints #1
Data point attributes:
     -> key: Str(value1)
StartTimestamp: 2024-01-12 08:50:25 +0000 UTC
Timestamp: 2024-01-12 08:50:30 +0000 UTC
Value: 1.000000
```
</details>

### CODE REVIEW

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java [hunk 1]
```diff
@@ -140,15 +131,10 @@ public void stop() {
 
     @Override
     protected void publish() {
-        if (isDelta()) {
```
Comment (marcingrzejszczak) [inline implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java:143]: Isn't this a breaking change?

**Review by marcingrzejszczak — APPROVED** [review]
WDYT @shakuzen @jonatan-ivanov ?

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java (hunk 1 above)
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMeterRegistry.java:143]: Nope. This is being moved to the OTLPMetricConverter([see here](https://github.com/micrometer-metrics/micrometer/blob/56b3607a68840fad0f6f3d5287f3f34cca8521a4/implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMetricConverter.java#L67)).

**Comment by lenin-jaganathan** [comment]
@marcingrzejszczak Did you get a chance to check back on this?

**Comment by shakuzen** [comment]
> 1. When a metric name has multiple descriptions (this shouldn't be an ideal usage but the micrometer allows this), we consider this a separate metric and add it to a new metric.

This has come up in the past. I think we should try to update all of our instrumentation to avoid having different descriptions. If we don't already have an issue to track this, feel free to open one, and we can try to tackle that in this upcoming minor.

**Comment by lenin-jaganathan** [comment]
I will write an follow-up to have more finer control of what should be defined as a meter.

File: implementations/micrometer-registry-otlp/src/test/java/io/micrometer/registry/otlp/OtlpMeterRegistryTest.java [hunk 2]
```diff
@@ -52,6 +52,11 @@ abstract class OtlpMeterRegistryTest {
 
     abstract OtlpConfig otlpConfig();
 
+    @AfterEach
+    void clearRegistry() {
+        registry.clear();
```
Comment (shakuzen) [inline implementations/micrometer-registry-otlp/src/test/java/io/micrometer/registry/otlp/OtlpMeterRegistryTest.java:57]: This shouldn't be necessary. Each test will run with a new instance of `OtlpMeterRegistry`

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMetricConverter.java [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+
+    OtlpMetricConverter(Clock clock, Duration step, TimeUnit baseTimeUnit,
+            AggregationTemporality aggregationTemporality, NamingConvention namingConvention) {
+        this.clock = clock;
+        this.step = step;
+        this.aggregationTemporality = aggregationTemporality;
+        this.otlpAggregationTemporality = AggregationTemporality.toOtlpAggregationTemporality(aggregationTemporality);
+        this.baseTimeUnit = baseTimeUnit;
+        this.namingConvention = namingConvention;
+        this.deltaTimeUnixNano = (clock.wallTime() / step.toMillis()) * step.toNanos();
```
Comment (shakuzen) [inline implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMetricConverter.java:67]: Isn't this (in combination with where the constructor is used in `publish`) changing from setting the delta time once per publish to setting it once per batch? I guess we don't have a test that would catch such a change.
Was the change intentional? It seems unrelated to the main purpose of the pull request.

**Review by shakuzen — APPROVED** [review]
Looks good overall. I left one question.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMetricConverter.java (hunk 3 above)
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMetricConverter.java:67]: Yeah, this was intentional to this per batch reason being, in normal scenarios where the meters are all read and marshaled into OTLP protocol completed within the step this would mean no change except that computation of delatTimeUnixNano happens multiple times (once per batch). NOTE: Irrespective of how many computations happen the value is the same.

In case, there are too many and all of them are not read within the same step interval (like half are read and some spill over), the value would have rolled over to the next step, and re-computing per batch would at least set the right timestamp for that step. But maybe the right thing would be to re-compute for every meter. That looks sub-optimal to me given that if this scenario happens already the metrics are flawed and users have to re-configure their steps to accommodate publishing all the meters. What do you think?

File: implementations/micrometer-registry-otlp/src/test/java/io/micrometer/registry/otlp/OtlpMeterRegistryTest.java (hunk 2 above)
Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/test/java/io/micrometer/registry/otlp/OtlpMeterRegistryTest.java]: removed it.

File: implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMetricConverter.java (hunk 3 above)
Comment (shakuzen) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMetricConverter.java:67]: I think it's fine to make the change, but it should be called out as a change in behavior that was made. This is both for review purposes and tracking/announcing changes to users. It's not clear from the title or description that this would be changed as part of this.

Comment (lenin-jaganathan) [reply implementations/micrometer-registry-otlp/src/main/java/io/micrometer/registry/otlp/OtlpMetricConverter.java:67]: I have also updated the description of PR to reflect the same.
